Add gomod package manager support

lcarva opened the pull request.

lcarva commented:
Unit tests are in the works!

lcarva commented on cachito/workers/pkg_manager.py:
```diff
@@ -0,0 +1,58 @@
+        cmd = ('go', 'list', '-m', 'all')
+
+        go_list = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                                   text=True, encoding='utf-8', env=env, cwd=source_dir)
```
Bummer, the `test` parameter was only introduced in python 3.7. It seems that the existing `universal_newlines` will do the same. I'll change it.

https://docs.python.org/3/library/subprocess.html#subprocess.Popen

lcarva replied:
Fixed!

mprahl commented on cachito/workers/pkg_manager.py:
```diff
@@ -0,0 +1,58 @@
+            if len(parts) == 1:
+                # This is the application itself, not a dependency
+                continue
+            if len(parts) > 2:
```
How about `if len(parts) != 2:`?

lcarva replied:
Actually, I like to separate the two use cases. In the case where there's only one part, we can just ignore, but we really want to warn if it unexpectedly has more than 2 parts.

mikebonnet commented on cachito/workers/pkg_manager.py:
```diff
@@ -0,0 +1,58 @@
+        }
+        cmd = ('go', 'list', '-m', 'all')
+
+        go_list = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
```
Could consider using `subprocess.run()` here instead.

lcarva replied:
Yeah. I was trying to keep consistency with the `scm.py` module. I'll make a separate commit to change them both over.

lcarva replied:
Done!

yashvardhannanavati commented on cachito/workers/pkg_manager.py:
```diff
@@ -0,0 +1,58 @@
+        return deps
+
+
+def _mount_app_src(archive_path, parent_dir):
```
or `_get_app_source`?

lcarva replied:
`_extract_app_src` is what I had stashed locally. Pushed now.

lcarva commented on cachito/workers/config.py:
```diff
@@ -22,6 +22,7 @@ class DevelopmentConfig(Config):
     """The development Cachito Celery configuration."""
 
     broker_url = 'amqp://cachito:cachito@rabbitmq:5672//'
+    athens_url = 'http://athens:3000'
```
Probably. Although I'm not sure if this should be considered validation or part of the startup process. Should we enhance the command 'wait-for-db' to also wait for athens? (With a different name)

I'm tempted to skip the validation at this point.

lcarva commented:
Hey all, I think I've addressed all your comments either by making code changes, or by replying to your message. Please have another looks when you get a chance. Removing WIP from title now.

mprahl commented on cachito/workers/pkg_manager.py:
```diff
@@ -0,0 +1,69 @@
+            log.error(
+                'Fetching gomod dependencies with "%s" failed with: %s',
+                ' '.join(cmd),
+                go_list.stdout,
```
The error is outputted to `stdout` and not `stderr`?

mprahl commented on cachito/workers/scm.py:
```diff
@@ -136,14 +134,13 @@ def clone_and_archive(self):
-            git_clone = subprocess.Popen(cmd, stderr=subprocess.PIPE)
-            _, error_output = git_clone.communicate()
+            git_clone = subprocess.run(cmd, capture_output=True, universal_newlines=True,
+                                       encoding='utf-8')
```
Optional: I'd prefer the following formatting:
```python
git_clone = subprocess.run(
    cmd, capture_output=True, universal_newlines=True, encoding='utf-8')
```

mikebonnet replied:
As counterpoint, I tend to use Lui's formatting. :)

mprahl commented on cachito/workers/scm.py:
```diff
@@ -158,16 +155,15 @@ def clone_and_archive(self):
-            git_archive = subprocess.Popen(cmd, stderr=subprocess.PIPE)
-            _, error_output = git_archive.communicate()
+            git_archive = subprocess.run(cmd, capture_output=True, universal_newlines=True,
+                                         encoding='utf-8')
```
Optional: Same as above

mprahl commented on tests/test_workers/test_pkg_manager.py:
```diff
@@ -0,0 +1,106 @@
+    golang.org/x/tools v0.0.0-20190325161752-5a8dccf5b48a
+    gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127
+    gopkg.in/yaml.v2 v2.2.2
+    """)
```
Optional I'd prefer if there wasn't an indentation here.
